runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/data/GrpcDataService.java

```diff
@@ -105,7 +105,9 @@ public void close() throws Exception {
         // Shutdown remaining clients
```

Comment:
There is a concurrent race between data() and close().

A BeamFnDataGrpcMultiplexer can get added to `additionalMultiplexers` during the close() call.
Thread 1: paused at `if (!connectedClient.set(multiplexer)) {`
Thread 2: closes multiplexers contained in `additionalMultiplexers`
Thread 1: adds new multiplexer to `additionalMultiplexers`

It turns out that we use queues in a bunch of places and it would be good to have a thread safe queue that can be closed/cancelled where adding to it would return false if the queue is in a terminal state and hence close() methods would be able to be guaranteed to have the correct clean up. We can defer this problem as it is pre-existing for now but a useful TODO/JIRA would help.

Reply:
Added https://issues.apache.org/jira/browse/BEAM-3811, noted on the field.